MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/QueryFeatureActivity.java

```diff
@@ -53,53 +78,233 @@ public void onMapReady(@NonNull final MapboxMap mapboxMap) {
+    GenerateViewIconTask(QueryFeatureActivity activity) {
+      this.activityRef = new WeakReference<>(activity);
+    }
+
+    @SuppressWarnings("WrongThread")
+    @Override
+    protected HashMap<String, Bitmap> doInBackground(FeatureCollection... params) {
+      QueryFeatureActivity activity = activityRef.get();
+      HashMap<String, Bitmap> imagesMap = new HashMap<>();
+      if (activity != null) {
+        LayoutInflater inflater = LayoutInflater.from(activity);
+
+        featureAtMapClickPoint = params[0].features().get(0);
+
+        StringBuilder stringBuilder = new StringBuilder();
+
+        BubbleLayout bubbleLayout = (BubbleLayout) inflater.inflate(
+          R.layout.activity_query_feature_window_symbol_layer, null);
+
+        TextView titleTextView = bubbleLayout.findViewById(R.id.info_window_title);
+        titleTextView.setText(activity.getString(R.string.query_feature_marker_title));
+
+        for (Map.Entry<String, JsonElement> entry : featureAtMapClickPoint.properties().entrySet()) {
           stringBuilder.append(String.format("%s - %s", entry.getKey(), entry.getValue()));
           stringBuilder.append(System.getProperty("line.separator"));
         }
 
-        featureMarker = mapboxMap.addMarker(new MarkerOptions()
-          .position(point)
-          .title(getString(R.string.query_feature_marker_title))
-          .snippet(stringBuilder.toString())
-        );
-
-      } else {
-        property = getString(R.string.query_feature_marker_snippet);
-        featureMarker = mapboxMap.addMarker(new MarkerOptions()
-          .position(point)
-          .snippet(property)
-        );
+        TextView propertiesListTextView = bubbleLayout.findViewById(R.id.info_window_feature_properties_list);
+        propertiesListTextView.setText(stringBuilder.toString());
+
+        int measureSpec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED);
+        bubbleLayout.measure(measureSpec, measureSpec);
+
+        int measuredWidth = bubbleLayout.getMeasuredWidth();
+
+        bubbleLayout.setArrowPosition(measuredWidth / 2 - 5);
+
+        Bitmap bitmap = QueryFeatureActivity.SymbolGenerator.generate(bubbleLayout);
+        imagesMap.put(CALLOUT_IMAGE_ID, bitmap);
+        viewMap.put(CALLOUT_IMAGE_ID, bubbleLayout);
       }
-    } else {
-      featureMarker = mapboxMap.addMarker(new MarkerOptions()
-        .position(point)
-        .snippet(getString(R.string.query_feature_marker_snippet))
-      );
+
+      return imagesMap;
     }
-    mapboxMap.selectMarker(featureMarker);
 
-    return true;
+    @Override
+    protected void onPostExecute(HashMap<String, Bitmap> bitmapHashMap) {
+      super.onPostExecute(bitmapHashMap);
+      QueryFeatureActivity activity = activityRef.get();
+      if (activity != null && bitmapHashMap != null) {
+        activity.setImageGenResults(bitmapHashMap);
+        activity.refreshSource(featureAtMapClickPoint);
```

Comment:
The source doesn't need to be refreshed at this point. Adding an image will update the map if necessary.

Reply:
Seems that it does need to be refreshed. When I remove `activity.refreshSource(featureAtMapClickPoint);`  and run the example, the property list info window icon doesn't appear when the map is tapped on.

Reply:
Ah, I thought it's also updated during the click itself.